File: cloud-flow-agent/internal/ebpfcollector/bpf/tcp_metrics_bpf.go
// tcp_metrics_bpf.go - Go绑定用于加载TCP指标eBPF程序
//go:build linux
// +build linux

package bpf

import (
	"bytes"
	"embed"
	"fmt"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"
)

//go:embed tcp_metrics.bpf.o
var tcpMetricsBpfFS embed.FS

// TCPMetricsObjects 包含所有TCP指标eBPF对象
type TCPMetricsObjects struct {
	// Programs
	TraceTcpV4ConnectEntry  *ebpf.Program `ebpf:"trace_tcp_v4_connect_entry"`
	TraceTcpV6ConnectEntry  *ebpf.Program `ebpf:"trace_tcp_v6_connect_entry"`
	TraceTcpRcvStateProcess *ebpf.Program `ebpf:"trace_tcp_rcv_state_process"`
	TraceTcpRetransmitSkb   *ebpf.Program `ebpf:"trace_tcp_retransmit_skb"`
	TraceTcpAckUpdateWindow *ebpf.Program `ebpf:"trace_tcp_ack_update_window"`
	TraceTcpV4SynRecvSock   *ebpf.Program `ebpf:"trace_tcp_v4_syn_recv_sock"`
	TraceTcpDrop            *ebpf.Program `ebpf:"trace_tcp_drop"`
	TraceTcpClose           *ebpf.Program `ebpf:"trace_tcp_close"`

	// Maps
	TcpLatencyMap        *ebpf.Map `ebpf:"tcp_latency_map"`
	TcpStatsMap          *ebpf.Map `ebpf:"tcp_stats_map"`
	GlobalTcpMetricsMap  *ebpf.Map `ebpf:"global_tcp_metrics_map"`
	ZeroWindowMap        *ebpf.Map `ebpf:"zero_window_map"`
	QueueOverflowMap     *ebpf.Map `ebpf:"queue_overflow_map"`
	ConnFailMap          *ebpf.Map `ebpf:"conn_fail_map"`
}

// TcpConnKey TCP连接标识
type TcpConnKey struct {
	Saddr uint32
	Daddr uint32
	Sport uint16
	Dport uint16
	Pid   uint32
}

// TcpLatency TCP连接时延数据
type TcpLatency struct {
	SynSentNs      uint64
	SynackRecvNs   uint64
	EstablishedNs  uint64
	LatencyNs      uint64
	Complete       uint8
	Padding        [7]byte
}

// TcpStats TCP统计指标
type TcpStats struct {
	RetransCount       uint64
	ZeroWindowCount    uint64
	QueueOverflowCount uint64
	ConnFailCount      uint64
	BytesSent          uint64
	BytesRecv          uint64
	PacketsSent        uint64
	PacketsRecv        uint64
	LastUpdate         uint64
}

// GlobalTcpMetrics 全局TCP指标汇总
type GlobalTcpMetrics struct {
	TotalConnections     uint64
	FailedConnections    uint64
	TotalRetrans         uint64
	ZeroWindowEvents     uint64
	QueueOverflowEvents  uint64
	AvgLatencyNs         uint64
	MaxLatencyNs         uint64
	MinLatencyNs         uint64
	LatencySamples       uint64
}

// LoadTCPMetrics 加载TCP指标eBPF程序
func LoadTCPMetrics(opts *ebpf.CollectionOptions) (*TCPMetricsObjects, error) {
	// 读取编译后的eBPF对象文件
	objData, err := tcpMetricsBpfFS.ReadFile("tcp_metrics.bpf.o")
	if err != nil {
		return nil, fmt.Errorf("读取TCP指标eBPF对象失败: %w", err)
	}

	spec, err := ebpf.LoadCollectionSpecFromReader(bytes.NewReader(objData))
	if err != nil {
		return nil, fmt.Errorf("加载TCP指标eBPF规格失败: %w", err)
	}

	var objs TCPMetricsObjects
	if err := spec.LoadAndAssign(&objs, opts); err != nil {
		return nil, fmt.Errorf("加载TCP指标eBPF对象失败: %w", err)
	}

	return &objs, nil
}

// Close 关闭所有eBPF对象
func (o *TCPMetricsObjects) Close() error {
	var errs []error
	
	if o.TraceTcpV4ConnectEntry != nil {
		if err := o.TraceTcpV4ConnectEntry.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.TraceTcpV6ConnectEntry != nil {
		if err := o.TraceTcpV6ConnectEntry.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.TraceTcpRcvStateProcess != nil {
		if err := o.TraceTcpRcvStateProcess.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.TraceTcpRetransmitSkb != nil {
		if err := o.TraceTcpRetransmitSkb.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.TraceTcpAckUpdateWindow != nil {
		if err := o.TraceTcpAckUpdateWindow.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.TraceTcpV4SynRecvSock != nil {
		if err := o.TraceTcpV4SynRecvSock.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.TraceTcpDrop != nil {
		if err := o.TraceTcpDrop.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.TraceTcpClose != nil {
		if err := o.TraceTcpClose.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	
	if o.TcpLatencyMap != nil {
		if err := o.TcpLatencyMap.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.TcpStatsMap != nil {
		if err := o.TcpStatsMap.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.GlobalTcpMetricsMap != nil {
		if err := o.GlobalTcpMetricsMap.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.ZeroWindowMap != nil {
		if err := o.ZeroWindowMap.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.QueueOverflowMap != nil {
		if err := o.QueueOverflowMap.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	if o.ConnFailMap != nil {
		if err := o.ConnFailMap.Close(); err != nil {
			errs = append(errs, err)
		}
	}
	
	if len(errs) > 0 {
		return fmt.Errorf("关闭TCP指标eBPF对象时发生错误: %v", errs)
	}
	return nil
}

// AttachTCPMetricsProbes 附加TCP指标kprobe探针
func AttachTCPMetricsProbes(objs *TCPMetricsObjects) ([]link.Link, error) {
	var links []link.Link
	
	// 附加tcp_v4_connect入口探针
	if objs.TraceTcpV4ConnectEntry != nil {
		l, err := link.Kprobe("tcp_v4_connect", objs.TraceTcpV4ConnectEntry, nil)
		if err != nil {
			return links, fmt.Errorf("附加tcp_v4_connect kprobe失败: %w", err)
		}
		links = append(links, l)
	}
	
	// 附加tcp_v6_connect入口探针
	if objs.TraceTcpV6ConnectEntry != nil {
		l, err := link.Kprobe("tcp_v6_connect", objs.TraceTcpV6ConnectEntry, nil)
		if err != nil {
			// IPv6可能不支持,不视为致命错误
			_ = err
		}
	}
	
	// 附加tcp_rcv_state_process探针
	if objs.TraceTcpRcvStateProcess != nil {
		l, err := link.Kprobe("tcp_rcv_state_process", objs.TraceTcpRcvStateProcess, nil)
		if err != nil {
			return links, fmt.Errorf("附加tcp_rcv_state_process kprobe失败: %w", err)
		}
		links = append(links, l)
	}
	
	// 附加tcp_retransmit_skb探针
	if objs.TraceTcpRetransmitSkb != nil {
		l, err := link.Kprobe("tcp_retransmit_skb", objs.TraceTcpRetransmitSkb, nil)
		if err != nil {
			return links, fmt.Errorf("附加tcp_retransmit_skb kprobe失败: %w", err)
		}
		links = append(links, l)
	}
	
	// 附加tcp_ack_update_window探针
	if objs.TraceTcpAckUpdateWindow != nil {
		l, err := link.Kprobe("tcp_ack_update_window", objs.TraceTcpAckUpdateWindow, nil)
		if err != nil {
			return links, fmt.Errorf("附加tcp_ack_update_window kprobe失败: %w", err)
		}
		links = append(links, l)
	}
	
	// 附加tcp_v4_syn_recv_sock探针
	if objs.TraceTcpV4SynRecvSock != nil {
		l, err := link.Kprobe("tcp_v4_syn_recv_sock", objs.TraceTcpV4SynRecvSock, nil)
		if err != nil {
			return links, fmt.Errorf("附加tcp_v4_syn_recv_sock kprobe失败: %w", err)
		}
		links = append(links, l)
	}
	
	// 附加tcp_drop探针
	if objs.TraceTcpDrop != nil {
		l, err := link.Kprobe("tcp_drop", objs.TraceTcpDrop, nil)
		if err != nil {
			// tcp_drop可能在某些内核中不可用
			_ = err
		} else {
			links = append(links, l)
		}
	}
	
	// 附加tcp_close探针
	if objs.TraceTcpClose != nil {
		l, err := link.Kprobe("tcp_close", objs.TraceTcpClose, nil)
		if err != nil {
			return links, fmt.Errorf("附加tcp_close kprobe失败: %w", err)
		}
		links = append(links, l)
	}
	
	return links, nil
}

// GetGlobalTCPMetrics 获取全局TCP指标
func (o *TCPMetricsObjects) GetGlobalTCPMetrics() (*GlobalTcpMetrics, error) {
	var key uint32 = 0
	var metrics GlobalTcpMetrics
	
	err := o.GlobalTcpMetricsMap.Lookup(&key, &metrics)
	if err != nil {
		return nil, fmt.Errorf("获取全局TCP指标失败: %w", err)
	}
	
	return &metrics, nil
}

// IterateTcpStats 遍历TCP统计映射
func (o *TCPMetricsObjects) IterateTcpStats() *ebpf.MapIterator {
	return o.TcpStatsMap.Iterate()
}

// IterateTcpLatency 遍历TCP时延映射
func (o *TCPMetricsObjects) IterateTcpLatency() *ebpf.MapIterator {
	return o.TcpLatencyMap.Iterate()
}

// IterateZeroWindow 遍历零窗口映射
func (o *TCPMetricsObjects) IterateZeroWindow() *ebpf.MapIterator {
	return o.ZeroWindowMap.Iterate()
}

// IterateQueueOverflow 遍历队列溢出映射
func (o *TCPMetricsObjects) IterateQueueOverflow() *ebpf.MapIterator {
	return o.QueueOverflowMap.Iterate()
}

// IterateConnFail 遍历连接失败映射
func (o *TCPMetricsObjects) IterateConnFail() *ebpf.MapIterator {
	return o.ConnFailMap.Iterate()
}

// ClearGlobalMetrics 清零全局指标
func (o *TCPMetricsObjects) ClearGlobalMetrics() error {
	var key uint32 = 0
	var empty GlobalTcpMetrics
	
	return o.GlobalTcpMetricsMap.Put(&key, &empty)
}

File: cloud-flow-agent/internal/ebpfcollector/bpf/tcp_metrics.bpf.c
// tcp_metrics.bpf.c - eBPF程序用于采集TCP深度指标
// 包括: TCP建连时延、重传率、零窗口事件、队列溢出、连接失败次数、吞吐量

#include "vmlinux.h"
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>
#include <bpf/bpf_core_read.h>

#define AF_INET 2
#define AF_INET6 10
#define TCP_ESTABLISHED 1
#define TCP_SYN_SENT 2
#define TCP_SYN_RECV 3
#define TCP_FIN_WAIT1 4
#define TCP_FIN_WAIT2 5
#define TCP_TIME_WAIT 6
#define TCP_CLOSE 7
#define TCP_CLOSE_WAIT 8
#define TCP_LAST_ACK 9
#define TCP_LISTEN 10
#define TCP_CLOSING 11

// TCP连接标识
struct tcp_conn_key {
    __u32 saddr;
    __u32 daddr;
    __u16 sport;
    __u16 dport;
    __u32 pid;
    __u8  comm[16];  // 进程名
};

// 统一的TCP流统计结构体 - 聚合所有指标
struct tcp_flow_stats {
    // 建连时延（端到端）
    __u64 connect_latency_ns;     // 建连时延(纳秒)
    __u64 syn_sent_ns;            // SYN 发送时间戳
    __u8  connect_complete;       // 连接是否已完成
    __u8  padding1[7];

    // 重传统计
    __u64 retrans_count;          // 重传次数

    // 零窗口事件
    __u64 zero_window_count;      // 零窗口事件次数

    // TCP 队列溢出
    __u64 queue_overflow_count;   // TCP 队列溢出次数

    // 连接失败
    __u64 conn_fail_count;        // 连接失败次数

    // 吞吐量
    __u64 bytes_sent;             // 发送字节数
    __u64 bytes_recv;             // 接收字节数
    __u64 packets_sent;           // 发送包数
    __u64 packets_recv;           // 接收包数

    // 元数据
    __u64 last_update;            // 最后更新时间
};

// 全局TCP指标汇总
struct global_tcp_metrics {
    __u64 total_connections;        // 总连接数
    __u64 failed_connections;       // 失败连接数
    __u64 total_retrans;            // 总重传数
    __u64 zero_window_events;       // 零窗口事件数
    __u64 queue_overflow_events;    // 队列溢出事件数
    __u64 avg_latency_ns;           // 平均建连时延
    __u64 max_latency_ns;           // 最大建连时延
    __u64 min_latency_ns;           // 最小建连时延
    __u64 latency_samples;          // 时延样本数
};

// BPF Maps定义

// 统一的TCP流统计Map（按连接维度聚合所有指标）
struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 65536);
    __type(key, struct tcp_conn_key);
    __type(value, struct tcp_flow_stats);
} tcp_flow_stats_map SEC(".maps");

// 全局TCP指标汇总Map
struct {
    __uint(type, BPF_MAP_TYPE_ARRAY);
    __uint(max_entries, 1);
    __type(key, __u32);
    __type(value, struct global_tcp_metrics);
} global_tcp_metrics_map SEC(".maps");

// 辅助函数: 获取当前时间戳(纳秒)
static __always_inline __u64 get_timestamp_ns(void) {
    return bpf_ktime_get_ns();
}

// 辅助函数: 从 sock 结构体中提取连接 key
static __always_inline void fill_conn_key(struct sock *sk, struct tcp_conn_key *key) {
    key->saddr = BPF_CORE_READ(sk, __sk_common.skc_rcv_saddr);
    key->daddr = BPF_CORE_READ(sk, __sk_common.skc_daddr);
    key->sport = BPF_CORE_READ(sk, __sk_common.skc_num);
    key->dport = bpf_ntohs(BPF_CORE_READ(sk, __sk_common.skc_dport));
    key->pid = bpf_get_current_pid_tgid() >> 32;
    bpf_get_current_comm(&key->comm, sizeof(key->comm));
}

// 辅助函数: 查找或创建 flow stats 条目
static __always_inline struct tcp_flow_stats *get_or_create_flow_stats(struct tcp_conn_key *key) {
    struct tcp_flow_stats *stats = bpf_map_lookup_elem(&tcp_flow_stats_map, key);
    if (!stats) {
        struct tcp_flow_stats init = {};
        bpf_map_update_elem(&tcp_flow_stats_map, key, &init, BPF_NOEXIST);
        stats = bpf_map_lookup_elem(&tcp_flow_stats_map, key);
    }
    return stats;
}

// ==================== TCP建连时延跟踪 ====================

// 跟踪tcp_v4_connect入口 - SYN发送
SEC("kprobe/tcp_v4_connect")
int BPF_KPROBE(trace_tcp_v4_connect_entry, struct sock *sk) {
    struct tcp_conn_key key = {};
    struct tcp_flow_stats init = {};

    // 读取连接信息并填充key
    fill_conn_key(sk, &key);

    // 初始化流统计条目，记录SYN发送时间
    init.syn_sent_ns = get_timestamp_ns();
    init.last_update = init.syn_sent_ns;

    bpf_map_update_elem(&tcp_flow_stats_map, &key, &init, BPF_ANY);

    return 0;
}

// 跟踪tcp_v6_connect入口
SEC("kprobe/tcp_v6_connect")
int BPF_KPROBE(trace_tcp_v6_connect_entry, struct sock *sk) {
    // IPv6逻辑类似,简化处理,只跟踪IPv4
    return 0;
}

// 跟踪tcp_rcv_state_process - 处理TCP状态变化
SEC("kprobe/tcp_rcv_state_process")
int BPF_KPROBE(trace_tcp_rcv_state_process, struct sock *sk) {
    struct tcp_conn_key key = {};
    struct tcp_flow_stats *stats;
    __u8 old_state;

    // 读取连接信息并填充key
    fill_conn_key(sk, &key);

    stats = bpf_map_lookup_elem(&tcp_flow_stats_map, &key);
    if (!stats)
        return 0;

    BPF_CORE_READ_INTO(&old_state, sk, __sk_common.skc_state);

    // 检测连接完全建立 (SYN_SENT -> ESTABLISHED 或 SYN_RECV -> ESTABLISHED)
    if ((old_state == TCP_SYN_SENT || old_state == TCP_SYN_RECV) &&
        stats->syn_sent_ns > 0 && !stats->connect_complete) {
        __u64 now = get_timestamp_ns();
        stats->connect_latency_ns = now - stats->syn_sent_ns;
        stats->connect_complete = 1;
        stats->last_update = now;

        // 更新全局时延统计
        __u32 gkey = 0;
        struct global_tcp_metrics *gmetrics = bpf_map_lookup_elem(&global_tcp_metrics_map, &gkey);
        if (gmetrics) {
            gmetrics->total_connections++;
            gmetrics->latency_samples++;
            // 使用近似平均计算
            gmetrics->avg_latency_ns =
                (gmetrics->avg_latency_ns * (gmetrics->latency_samples - 1) + stats->connect_latency_ns)
                / gmetrics->latency_samples;
            if (stats->connect_latency_ns > gmetrics->max_latency_ns)
                gmetrics->max_latency_ns = stats->connect_latency_ns;
            if (gmetrics->min_latency_ns == 0 || stats->connect_latency_ns < gmetrics->min_latency_ns)
                gmetrics->min_latency_ns = stats->connect_latency_ns;
        }
    }

    return 0;
}

// ==================== TCP重传跟踪 ====================

// 跟踪tcp_retransmit_skb - TCP重传
SEC("kprobe/tcp_retransmit_skb")
int BPF_KPROBE(trace_tcp_retransmit_skb, struct sock *sk, struct sk_buff *skb) {
    struct tcp_conn_key key = {};
    struct tcp_flow_stats *stats;

    fill_conn_key(sk, &key);

    stats = get_or_create_flow_stats(&key);
    if (stats) {
        stats->retrans_count++;
        stats->last_update = get_timestamp_ns();
    }

    // 更新全局重传计数
    __u32 gkey = 0;
    struct global_tcp_metrics *gmetrics = bpf_map_lookup_elem(&global_tcp_metrics_map, &gkey);
    if (gmetrics) {
        gmetrics->total_retrans++;
    }

    return 0;
}

// ==================== 零窗口事件跟踪 ====================

// 跟踪tcp_ack_update_window - 检测零窗口通告
SEC("kprobe/tcp_ack_update_window")
int BPF_KPROBE(trace_tcp_ack_update_window, struct sock *sk, __u32 nwin) {
    // nwin为0表示零窗口
    if (nwin != 0)
        return 0;

    struct tcp_conn_key key = {};
    struct tcp_flow_stats *stats;

    fill_conn_key(sk, &key);

    stats = get_or_create_flow_stats(&key);
    if (stats) {
        stats->zero_window_count++;
        stats->last_update = get_timestamp_ns();
    }

    // 更新全局零窗口计数
    __u32 gkey = 0;
    struct global_tcp_metrics *gmetrics = bpf_map_lookup_elem(&global_tcp_metrics_map, &gkey);
    if (gmetrics) {
        gmetrics->zero_window_events++;
    }

    return 0;
}

// ==================== 队列溢出跟踪 ====================

// 跟踪tcp_v4_syn_recv_sock - SYN队列处理
SEC("kprobe/tcp_v4_syn_recv_sock")
int BPF_KPROBE(trace_tcp_v4_syn_recv_sock, struct sock *sk, struct sk_buff *skb,
               struct request_sock *req, struct dst_entry *dst,
               struct request_sock *req_unhash, bool *own_req) {
    // 如果req为NULL,表示队列已满
    if (!req)
        return 0;

    struct tcp_conn_key key = {};
    struct tcp_flow_stats *stats;

    fill_conn_key(sk, &key);

    stats = get_or_create_flow_stats(&key);
    if (stats) {
        stats->queue_overflow_count++;
        stats->last_update = get_timestamp_ns();
    }

    // 更新全局队列溢出计数
    __u32 gkey = 0;
    struct global_tcp_metrics *gmetrics = bpf_map_lookup_elem(&global_tcp_metrics_map, &gkey);
    if (gmetrics) {
        gmetrics->queue_overflow_events++;
    }

    return 0;
}

// ==================== 连接失败跟踪 ====================

// 跟踪tcp_drop - 连接被丢弃
SEC("kprobe/tcp_drop")
int BPF_KPROBE(trace_tcp_drop, struct sock *sk, struct sk_buff *skb) {
    struct tcp_conn_key key = {};
    struct tcp_flow_stats *stats;

    fill_conn_key(sk, &key);

    stats = get_or_create_flow_stats(&key);
    if (stats) {
        stats->conn_fail_count++;
        stats->last_update = get_timestamp_ns();
    }

    // 更新全局失败计数
    __u32 gkey = 0;
    struct global_tcp_metrics *gmetrics = bpf_map_lookup_elem(&global_tcp_metrics_map, &gkey);
    if (gmetrics) {
        gmetrics->failed_connections++;
    }

    return 0;
}

// ==================== TCP吞吐量跟踪 ====================

// 跟踪tcp_sendmsg - 发送数据
SEC("kprobe/tcp_sendmsg")
int BPF_PROG(trace_tcp_sendmsg, struct sock *sk, struct msghdr *msg, size_t size) {
    struct tcp_conn_key key = {};
    struct tcp_flow_stats *stats;

    fill_conn_key(sk, &key);

    stats = get_or_create_flow_stats(&key);
    if (stats && size > 0) {
        stats->bytes_sent += size;
        stats->packets_sent++;
        stats->last_update = get_timestamp_ns();
    }

    return 0;
}

// 跟踪tcp_recvmsg - 接收数据
SEC("kprobe/tcp_recvmsg")
int BPF_PROG(trace_tcp_recvmsg, struct sock *sk, struct msghdr *msg, size_t len, int flags) {
    struct tcp_conn_key key = {};
    struct tcp_flow_stats *stats;

    fill_conn_key(sk, &key);

    stats = get_or_create_flow_stats(&key);
    if (stats && len > 0) {
        // recvmsg入口时len为请求读取的大小，实际接收量在retprobe中获取
        // 此处记录请求大小的近似值用于流量估算
        stats->bytes_recv += len;
        stats->packets_recv++;
        stats->last_update = get_timestamp_ns();
    }

    return 0;
}

// ==================== TCP连接关闭清理 ====================

// 跟踪tcp_close - 连接关闭时清理
SEC("kprobe/tcp_close")
int BPF_KPROBE(trace_tcp_close, struct sock *sk, long timeout) {
    struct tcp_conn_key key = {};

    fill_conn_key(sk, &key);

    // 清理流统计映射
    bpf_map_delete_elem(&tcp_flow_stats_map, &key);

    return 0;
}

// 许可证声明
char LICENSE[] SEC("license") = "GPL";
